src/openforms/tests/e2e/test_input_validation.py

```diff
@@ -670,3 +681,133 @@ def test_regex_pattern(self):
             ui_input="<h2>test</h2>",
             expected_ui_error="Ongeldig Nederlands kenteken",
         )
+
+
+class SingleFileTests(ValidationsTestCase):
+
+    def assertFileValidationIsAligned(
+        self,
+        component: Component,
+        ui_files: list[Path],
+        expected_ui_error: str,
+        api_value: list[dict[str, Any]],
+    ) -> None:
+        form = create_form(component)
+
+        with self.subTest("frontend validation"):
+            self._assertFileFrontendValidation(form, ui_files, expected_ui_error)
+
+        with self.subTest("backend validation"):
+            self._assertBackendValidation(form, component["key"], api_value)
+
+    @async_to_sync
+    async def _assertFileFrontendValidation(
+        self, form: Form, ui_files: list[Path], expected_ui_error: str
+    ) -> None:
+        frontend_path = reverse("forms:form-detail", kwargs={"slug": form.slug})
+        url = str(furl(self.live_server_url) / frontend_path)
+
+        async with browser_page() as page:
+            await page.goto(url)
+            await page.get_by_role("button", name="Formulier starten").click()
+
+            async with page.expect_file_chooser() as fc_info:
+                await page.get_by_text("blader").click()
+
+            file_chooser = await fc_info.value
+            await file_chooser.set_files(ui_files)
+            await page.wait_for_load_state("networkidle")
+
+            # try to submit the step which should be invalid, so we expect this to
+            # render the error message.
+            await page.get_by_role("button", name="Volgende").click()
+            await expect(page.get_by_text(expected_ui_error)).to_be_visible()
+
+    def test_required_field(self):
+        component = {
+            "type": "file",
+            "key": "requiredFile",
+            "label": "Required file",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        self.assertFileValidationIsAligned(
+            component,
+            ui_files=[],
+            expected_ui_error="Het verplichte veld Required file is niet ingevuld.",
+            api_value=[],
+        )
+
+    def test_bad_pdf(self):
+        component = {
+            "type": "file",
+            "key": "badPdf",
+            "label": "Bad PDF",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        # The frontend validation will *not* create a TemporaryFileUpload,
+        # as the endpoint will return a 400 because of the bad content type.
+        # For this reason, we manually create an invalid TemporaryFileUpload
+        # and use it for the `api_value`:
+        with open(TEST_FILES / "image-256x256.pdf", "rb") as infile:
+            temporary_upload = TemporaryFileUploadFactory.create(
+                file_name="image-256x256.pdf",
+                content=File(infile),
+                content_type="application/pdf",
+            )
```

Comment:
This is not realistic - because it's the endpoint itself that blocks the upload (or is supposed to), there is no way that a user can get this mismatched content type uploaded to the backend.

Instead, I would drop this bit and at the end of the test make an assertion that no temporary file uploads exist at all. For the upload itself (`api_value`), you can just generate a random UUID and the backend validation is then expected to report that no such temporary upload exists.